### ISSUE TITLE
Error when using generate-lines with CPU model 

### ISSUE THREAD

**Issue by shahbazsyed** [facebookresearch/fairseq-lua#17]
Hi,
I am trying to test the translation for the English to German CPU model. I get the following error when trying to load the model

![image](https://cloud.githubusercontent.com/assets/4356098/25941913/63777618-363b-11e7-9d57-38c92108c0b7.png)

**Comment by jgehring** [comment]
Hi @shahbazsyed, try updating your nn installation, e.g. via `luarocks install nn`.

**Comment by ddurham2** [comment]
I'm having the same problem.  However when I run the "luarocks install nn", it's giving a compile error which may be due to my environment, but I've only done what the torch installation instructions said to do...
`
.....

-- Generating done                                                                                                                                                                                  
-- Build files have been written to: /home/ddurham/tmp/luarocks_nn-scm-1-7697/nn/build                                                                                                              
Scanning dependencies of target THNN                                                                                                                                                                
[ 50%] Building C object lib/THNN/CMakeFiles/THNN.dir/init.c.o                                                                                                                                      
In file included from /home/ddurham/torch/install/include/TH/TH.h:4:0,                                                                                                                              
                 from /home/ddurham/tmp/luarocks_nn-scm-1-7697/nn/lib/THNN/init.c:1:                                                                                                                
/home/ddurham/tmp/luarocks_nn-scm-1-7697/nn/lib/THNN/generic/SparseLinear.c: In function ‘THNN_FloatSparseLinear_accGradParameters’:                                                                
/home/ddurham/torch/install/include/TH/THTensor.h:8:39: error: too many arguments to function ‘THFloatTensor_sum’                                                                                   
 #define THTensor_(NAME)   TH_CONCAT_4(TH,Real,Tensor_,NAME)                                                                                                                                        
                                       ^                                                                                                                                                            
/home/ddurham/torch/install/include/TH/THGeneral.h:108:37: note: in definition of macro ‘TH_CONCAT_4_EXPAND’                                                                                        
 #define TH_CONCAT_4_EXPAND(x,y,z,w) x ## y ## z ## w                                                                                                                                               
                                     ^                                                                                                                                                              
/home/ddurham/torch/install/include/TH/THTensor.h:8:27: note: in expansion of macro ‘TH_CONCAT_4’                                                                                                   
 #define THTensor_(NAME)   TH_CONCAT_4(TH,Real,Tensor_,NAME)                                                                                                                                        
                           ^                                                                                                                                                                        
/home/ddurham/tmp/luarocks_nn-scm-1-7697/nn/lib/THNN/generic/SparseLinear.c:237:3: note: in expansion of macro ‘THTensor_’                                                                          
   THTensor_(sum)(buf, gradOutput, 0, 1);                                                                                                                                                           
   ^                                                                                                                                                                                                
/home/ddurham/torch/install/include/TH/THTensor.h:8:39: note: declared here                                                                                                                         
 #define THTensor_(NAME)   TH_CONCAT_4(TH,Real,Tensor_,NAME)                                                                                                                                        
                                       ^                                                                                                                                                            
/home/ddurham/torch/install/include/TH/THGeneral.h:108:37: note: in definition of macro ‘TH_CONCAT_4_EXPAND’                                                                                        
 #define TH_CONCAT_4_EXPAND(x,y,z,w) x ## y ## z ## w                                                                                                                                               
                                     ^                                                                                                                                                              
/home/ddurham/torch/install/include/TH/THTensor.h:8:27: note: in expansion of macro ‘TH_CONCAT_4’                                                                                                   
 #define THTensor_(NAME)   TH_CONCAT_4(TH,Real,Tensor_,NAME)                                                                                                                                        
                           ^                                                                                                                                                                        
/home/ddurham/torch/install/include/TH/generic/THTensorMath.h:77:13: note: in expansion of macro ‘THTensor_’                                                                                        
 TH_API void THTensor_(sum)(THTensor *r_, THTensor *t, int dimension);                                                                                                                              
             ^                                                                                                                              
....
 `
other similar errors follow.  Ever run into this?

**Comment by mainka** [comment]
Same here. Fresh Torch install on Ubuntu 16 intel64. Unable to upgrade nn with the same error as @ddurham2.

**Comment by jgehring** [comment]
@ddurham2 and @mainka, I don't really know what's wrong here. I suggest heading over to https://github.com/torch/nn for searching prior issues and then filing a new one if necessary.

**Comment by shahbazsyed** [comment]
@ddurham2 @mainka Try doing a luarocks install torch once before luarocks install nn . Worked for me. I am not sure what was wrong as I also did a fresh install of torch before starting with fairseq. 

Anyhow, hope it helps. I could load the CPU model.

**Comment by shahbazsyed** [comment]
@jgehring Is there a cmdline argument to specify the gpuid to use for loading the CUDA model for testing? I have 8 GPUs(8GB memory each) and 6 of them are in use by other process. I don't understand how the model is being loaded but it always gives me out of memory error after loading the dictionaries though I have 2 GPUs free.

If it helps, I am running this inside a docker container on the host with 8 GPUs.

**Comment by jgehring** [comment]
No there is no such flag; I suggest using the [CUDA_VISIBLE_DEVICES](https://devblogs.nvidia.com/parallelforall/cuda-pro-tip-control-gpu-visibility-cuda_visible_devices/) environment variable to control which GPUs are available to the program.

**Comment by jgehring** [comment]
Please see the discussion at #18

**Comment by ddurham2** [comment]
Yes it's working for me.  Over at nn, another issue that installing torch from luarocks works.  I think nn simply has a problem compiling against the HEAD of torch (i.e. when you follow their directions of cloning from git and building).

**Comment by jgehring** [comment]
Ok, thanks for the context!

**Comment by WilliamTambellini** [comment]
Got the same error: looks like torch nn does not compile anymore with torch master:
luarocks install nn
Installing https://raw.githubusercontent.com/torch/rocks/master/nn-scm-1.rockspec...
Using https://raw.githubusercontent.com/torch/rocks/master/nn-scm-1.rockspec... switching to 'build' mode
Cloning into 'nn'...
remote: Counting objects: 319, done.
remote: Compressing objects: 100% (278/278), done.
remote: Total 319 (delta 76), reused 140 (delta 38), pack-reused 0
Receiving objects: 100% (319/319), 601.91 KiB | 0 bytes/s, done.
Resolving deltas: 100% (76/76), done.
Checking connectivity... done.
cmake -E make_directory build && cd build && cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH="/home/wtambellini/torch/install/bin/.." -DCMAKE_INSTALL_PREFIX="/home/wtambellini/torch/install/lib/luarocks/rocks/nn/scm-1" && make

-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Torch7 in /home/wtambellini/torch/install
-- TH_LIBRARIES: TH
-- Try OpenMP C flag = [-fopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Success
-- Try OpenMP CXX flag = [-fopenmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Success
-- Found OpenMP: -fopenmp  
-- Compiling with OpenMP support
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/luarocks_nn-scm-1-9722/nn/build
Scanning dependencies of target THNN
[ 50%] Building C object lib/THNN/CMakeFiles/THNN.dir/init.c.o
In file included from /home/wtambellini/torch/install/include/TH/TH.h:4:0,
                 from /tmp/luarocks_nn-scm-1-9722/nn/lib/THNN/init.c:1:
/tmp/luarocks_nn-scm-1-9722/nn/lib/THNN/generic/SparseLinear.c: In function ‘THNN_FloatSparseLinear_accGradParameters’:
/home/wtambellini/torch/install/include/TH/THTensor.h:8:39: error: too many arguments to function ‘THFloatTensor_sum’
 #define THTensor_(NAME)   TH_CONCAT_4(TH,Real,Tensor_,NAME)                                       ^
/home/wtambellini/torch/install/include/TH/THGeneral.h:108:37: note: in definition of macro ‘TH_CONCAT_4_EXPAND’
 #define TH_CONCAT_4_EXPAND(x,y,z,w) x ## y ## z ## w

even if I do a 
luarocks install torch 
before
luarocks install nn
Cheers
W.
